Add Stream Filter tests from Functional Test Suite.

jthelin opened the pull request:
Add StreamFilteringTests to GitHub test suite.

- Add StreamFilteringTests test cases from Orleans Functional test suite.

- Use FluentAssertions package for easier way to express expected outcomes in test cases.

- Changed default namespaces set in TestGrains.csproj and TestGrainInterfaces.csproj to match the current naming convention we use.

jthelin commented:
Is anyone able to repo the test failures that Jenkins gives for this PR / branch?

This branch builds & tests ok for me when run locally and on my Appveyer CI builds, so i am not sure what is sending Jenkins into a funk today. :(

veikkoeeva commented:
@jthelin I'll give it a run in about ten hours to see if I have a failure. If the problem persists that is.

jthelin commented:
Rebased against master 99bead4d812a55b78e8cb7614200196c6098915f to fix merge conflicts.

jthelin commented:
For comparison, here is the Appveyer build of this PR branch at 125cae72d7c0592ee0d41c25623f679ec0d057b2

https://ci.appveyor.com/project/jthelin/orleans/build/1.0.348/tests

veikkoeeva commented:
@jthelin Hmm, meant in about 20 hours. :) Should the console show stdout? It looks like tere was a fairly recent issue in [Jenkins Jira](https://issues.jenkins-ci.org/browse/JENKINS-17506) to show MSTest output. The implementation looks like being [this](http://twocentsonsoftware.blogspot.fi/2012/10/getting-your-mstest-unit-test-stdout-to.html). I wonder would they help? At work currently, so hard to check properly, just splashing around before getting to check the branch on my machine.

veikkoeeva commented:
@jthelin The tests pass on my machine, so no help from here. I note they feel very slow and on the console output too there is a preceding timeout before an avalanche of failures. I'm too tired (:)) to reopen the output, but I wonder if they belong to the same group or something and fail because of the timeout until a next batch is started?

jthelin commented:
Thanks @veikkoeeva -- i'll dig a little deeper.

sergeybykov commented:
Don't these tests have a dependency on Azure Storage Emulator or Storage Service? Jenkins doesn't have the emulator.

jthelin commented:
None of the Azure test cases are in the BVT suite.
https://github.com/jthelin/orleans/commit/6997b279cdfdf02aa0de03f73862dedc1ad83648

jthelin commented:
Well, Jenkins build still fails after I tried using `vstest.console.exe` for test engine (which both Appveyer and VSO use) instead of `MsTest.exe`.

Failure mode is slightly different with VsTest, but still very puzzling:

```
Failed   SMS_Filter_Basic
Error Message:
   Test 'SMS_Filter_Basic' exceeded execution timeout period.

Failed   SMS_Filter_EvenOdd
Error Message:
   Class Initialization method Tester.StreamingTests.StreamFilteringTests.ClassInitialize threw exception. System.Threading.ThreadAbortException: System.Threading.ThreadAbortException: Thread was being aborted..
Stack Trace:
    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.VisualStudio.TestTools.TestTypes.Unit.UnitTestExecuter.RunClassInitializeMethod()
```

Paging Sherlock Holmes!

jthelin commented:
Well, that last experimental commit proved to be very interesting! https://github.com/dotnet/orleans/commit/01992b9a88731fca3f712da9fe844c93438ce4e6
BVT tests passed ok this time :)

I split the SMS and AQ test cases into separate test classes, and removed the AQ providers from the silo config used in the SMS tests.

Except for initializing the AQ stream provider during silo init in the prev ious combined case, nothing else related to AQ is happening during the SMS test cases, so is that really a root cause?
More thought required, I think.

jthelin commented:
Paging @sergeybykov -- what do you think we should do with this PR?

sergeybykov commented:
@jthelin I need to look into this. I have a vendor (Ningna) now who started working on migrating all of the functional tests to GitHub last week. We need to look if this overlaps with what she is doing (likely not) or if this can inform some of her work.

jthelin commented:
Squashed commits and rebased against latest master HEAD @ 52975575a1a10ba95b323b15f2c78b17e7e441c2

jthelin commented:
@sergeybykov -- I turned off the AQ test cases here, which seems to be what is causing the earlier Jenkins build problem in this PR.

I think we can merge this PR to master now, if you are OK?

I really do think we need to get some test coverage for stream filter functionality in the code GH test suites (I was actually surprise we had none in GH already), so I strongly recommend to merge this soon (even though it only exercises SMS side of the house)

Then we can circle back on the AQ tests separately (those still seem to be working OK in the internal test suite, so it may just be a Azure Storage Emulator vs Azure Cloud thing?) -- and might also be related to some weird Azure problems we hit while cleaning up the Geo changes for submission.

AFAICT there is nothing particularly special about the AQ test cases for stream filtering compared to other SMS+AQ tests we run, so the earlier Jenkins build problems are still quite puzzling -- especially when tests work fine in AppVeyor and local runs.

CC; @NingnaD

sergeybykov commented:
Now I'm having 'fun' integrating this back to VSO because of the missing grain methods with internal access.

Looking at that, I'm even more convinced now that the more productive approach is to add projects for grain interfaces and classes with internal access to the runtime for the tests that currently rely on that, and *then* work on refactoring them to eliminate such a need where possible. If we take the purist approach of "no test grains should ever have internal access to the runtime", it'll take much longer to bring the tests over.

gabikliot commented:
Don't you just delete the old tests from vso? Why do you need to "integrate" back?

sergeybykov commented:
Because there are also other tests there that were using the same test grain interfaces/classes but with additional methods that require internal access.

gabikliot commented:
Of course. :-(
A good strategy to handle that is to use a different namespaces for VSO and GH tests/grain/interfaces (and maybe also different prefix for GH test names) and then you can go gradually by eliminating the VSO tests, that can co-exist with GH for a while.

Anyway, you probably don't need my advise in that regard.
